server/readonly/readonly_service.ts

```diff
@@ -0,0 +1,113 @@
+import { IAuthenticationType } from '../auth/types/authentication_type';
+import { SecuritySessionCookie } from '../session/security_cookie';
+import { SecurityPluginConfigType } from '../index';
+
+export class ReadonlyService implements IReadOnlyService {
+  private readonly logger: Logger;
+  private readonly securityClient: SecurityClient;
+  private readonly auth: IAuthenticationType;
+  private readonly securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>;
+  private readonly config: SecurityPluginConfigType;
+
+  constructor(
+    logger: Logger,
+    securityClient: SecurityClient,
+    auth: IAuthenticationType,
+    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    config: SecurityPluginConfigType
+  ) {
+    this.logger = logger;
+    this.securityClient = securityClient;
+    this.auth = auth;
+    this.securitySessionStorageFactory = securitySessionStorageFactory;
+    this.config = config;
+  }
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    try {
+      const url = new URL(request.headers.referer as string);
+      const pathsToIgnore = ['login', 'logout', 'customerror'];
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
+    }
+  }
+
+  isReadOnlyTenant(authInfo: any): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
+
+    // private tenant is not affected
+    if (isPrivateTenant(currentTenant)) {
+      return false;
+    }
+
+    return authInfo.tenants[currentTenant] !== true;
+  }
+
+  async isReadonly(request: OpenSearchDashboardsRequest): Promise<boolean> {
+    // omit for anonymous pages to avoid authentication errors
+    if (this.isAnonymousPage(request)) {
+      return false;
+    }
+
+    if (!this.config.multitenancy.enabled) {
+      return false;
+    }
+
+    try {
+      const cookie = await this.securitySessionStorageFactory.asScoped(request).get();
+      let headers = request.headers;
+
+      if (!this.auth.requestIncludesAuthInfo(request) && cookie) {
+        headers = this.auth.buildAuthHeaderFromCookie(cookie, request);
+      }
+
+      const authInfo = await this.securityClient.authinfo(request, headers);
+
+      if (!authInfo.user_requested_tenant && cookie) {
+        authInfo.user_requested_tenant = cookie.tenant;
+      }
+
+      return authInfo && this.isReadOnlyTenant(authInfo);
+    } catch (error: any) {
+      return false;
+    }
+  }
+
+  async hideForReadonly(
+    request: OpenSearchDashboardsRequest,
+    capabilites: Capabilities,
+    hideCapabilities: Partial<Capabilities>
+  ): Promise<Partial<Capabilities>> {
+    return (await this.isReadonly(request)) ? merge(capabilites, hideCapabilities) : capabilites;
+  }
+}
```

Comment:
Could we add tests to verify this change?

Reply:
Sure, gonna do it